File: backend/src/PMS.Domain/Entities/SavedView.cs
using PMS.Domain.Common;
using PMS.Domain.Enums;

namespace PMS.Domain.Entities;

/// <summary>
/// Một góc nhìn đã lưu trên danh sách task của MỘT project (ADR-061) — bộ lọc + sắp xếp +
/// gom nhóm + tập cột hiển thị.
///
/// <para>
/// Đây là mảnh cuối của "nền tảng mở rộng": ADR-052 cho đội tự khai <b>cột</b>, ADR-059 cho
/// tự khai <b>trường</b>, ADR-060 cho tự khai <b>loại việc</b> — còn đây cho họ tự dựng
/// <b>góc nhìn</b> thay vì nhận một bố cục cố định.
/// </para>
/// <para>
/// 🔑 <b>Và nó là tiền đề thật của tầng "bộ máy quy trình" (§0).</b> Một "hàng đợi" —
/// <i>"mọi Change Request đang chờ tôi duyệt"</i> — chính là một view lưu được, không phải
/// một màn hình mới. Làm quy trình trước sẽ phải dựng một màn danh sách tạm rồi vứt đi.
/// </para>
/// </summary>
public class SavedView : BaseEntity
{
    public Guid ProjectId { get; set; }
    public Project Project { get; set; } = null!;

    /// <summary>Tên do người dùng đặt — "Việc quá hạn của tôi", "CR chờ duyệt"…</summary>
    public string Name { get; set; } = string.Empty;

    /// <summary>
    /// Người tạo. Giữ lại cả khi view đã <see cref="IsShared"/>: nó là thông tin quy trách
    /// nhiệm ("ai dựng cái này"), không chỉ là khoá phân quyền.
    /// </summary>
    public Guid OwnerId { get; set; }
    public Employee Owner { get; set; } = null!;

    /// <summary>
    /// <c>false</c> = chỉ chủ sở hữu thấy; <c>true</c> = mọi thành viên project thấy.
    ///
    /// <para>
    /// ⚠️ Quyền GHI lên một view đã chia sẻ là <b>chủ sở hữu HOẶC người có
    /// <c>ManageSavedViews</c></b> (PM), không phải mọi thành viên. Chỉ-chủ-sở-hữu sẽ khoá
    /// chết view chung của cả đội khi người đó rời dự án; mọi-thành-viên thì ai cũng ghi đè
    /// công của nhau. Xem ADR-061.
    /// </para>
    /// </summary>
    public bool IsShared { get; set; }

    /// <summary>Thứ tự hiển thị trên thanh chọn view. Không unique — xem <see cref="BoardColumn.Order"/>.</summary>
    public int Order { get; set; }

    // ---------- sắp xếp ----------

    /// <summary>
    /// Trường dựng sẵn dùng để sắp xếp. Rỗng = mặc định (hạn rồi tie-break bằng Id).
    ///
    /// <para>
    /// 📌 <b>Cố ý CHỈ trường dựng sẵn — sắp theo trường tuỳ biến chưa ship, và lý do là
    /// cascade chứ không phải độ khó.</b> Một khoá ngoại <c>SortByFieldDefinitionId</c> trên
    /// chính hàng này buộc phải là <c>SET NULL</c> (xoá trường thì view mất kiểu sắp xếp chứ
    /// không được biến mất theo). Nhưng khi đó <c>FieldDefinitions</c> chạm tới
    /// <see cref="SavedViewFilter"/> bằng <b>hai</b> lối —
    /// <c>FieldDefinitions →SET NULL→ SavedViews →CASCADE→ SavedViewFilters</c> và
    /// <c>FieldDefinitions →CASCADE→ SavedViewFilters</c> — mà phép kiểm của SQL Server là
    /// <b>tĩnh</b>, không quan tâm lối thứ hai có bao giờ chạy hay không. Đó đúng là hình
    /// dạng đã làm 18 test đỏ trong 12ms ở ADR-059.
    /// </para>
    /// <para>
    /// Lối thoát nếu về sau cần: một bảng <c>SavedViewSorts</c> riêng cùng khuôn
    /// <see cref="SavedViewFilter"/> (khi đó cả hai lối đều là CASCADE vào một bảng con
    /// không nằm dưới bảng kia — an toàn, đúng phân tích của ADR-060). Chưa làm vì nó là một
    /// bảng cho tối đa một hàng. <b>Lọc</b> theo trường tuỳ biến — phần thật sự đáng giá, và
    /// là thứ cột-có-kiểu của ADR-059 được mua về để phục vụ — thì hoạt động đầy đủ.
    /// </para>
    /// </summary>
    public TaskField? SortBy { get; set; }

    public bool SortDescending { get; set; }

    // ---------- gom nhóm ----------

    /// <summary>
    /// Gom nhóm — <b>chỉ trường dựng sẵn</b>, cố ý.
    ///
    /// <para>
    /// 📌 Gom theo một trường <c>MultiSelect</c> thì một task thuộc NHIỀU nhóm cùng lúc, và
    /// khi đó tổng số dòng hiển thị lớn hơn số task — một màn hình tự mâu thuẫn mà người
    /// dùng không có cách nào hiểu. Chưa có câu trả lời tốt nên chưa ship (§0, nguyên tắc
    /// "không ship một cờ không chặn được gì").
    /// </para>
    /// </summary>
    public TaskField? GroupBy { get; set; }

    // ---------- thành phần ----------

    public ICollection<SavedViewFilter> Filters { get; set; } = new List<SavedViewFilter>();

    /// <summary>
    /// Tập cột hiển thị, theo thứ tự. Rỗng = dùng bộ cột mặc định của màn danh sách — cố ý
    /// KHÔNG nhân bản danh sách mặc định vào từng view: làm vậy thì đổi bộ mặc định về sau
    /// sẽ không chạm tới view nào đã tạo, và không ai hiểu vì sao.
    /// </summary>
    public ICollection<SavedViewColumn> Columns { get; set; } = new List<SavedViewColumn>();
}

File: frontend/types/saved-view.ts
import type { FieldOptionResponse, FieldType } from './custom-field';
import type { TaskSummaryResponse } from './task';

/**
 * View lưu được trên danh sách task (ADR-061) — mảnh cuối của nền tảng mở rộng.
 *
 * 🔴 Hai mức quyền KHÁC nhau, đừng gộp khi gác UI:
 *  - View **RIÊNG**: ai cũng tạo/sửa/xoá của chính mình, kể cả `Viewer` — nó chỉ đổi cách
 *    họ nhìn danh sách, không ai khác thấy (cùng lý lẽ "theo dõi task" của ADR-036).
 *  - View **CHIA SẺ**: cần quyền PM. Đây là cấu hình cả đội nhìn thấy.
 *
 * Đừng tự suy quyền sửa ở frontend — backend trả sẵn `canEdit` trên từng view, vì luật là
 * "chủ sở hữu HOẶC PM" mà vế thứ hai frontend chỉ biết qua vai trò. Hai nơi cùng dựng một
 * luật thì chắc chắn có lúc lệch (ADR-034).
 */

/**
 * Trường **dựng sẵn** lọc/sắp xếp/hiển thị được. Danh mục ĐÓNG — backend từ chối giá trị
 * ngoài danh sách này.
 *
 * 📌 Cố ý chưa có `Labels`/`Watchers`: quan hệ N–N nên "bằng" có hai nghĩa (chứa nhãn này /
 * có đúng tập nhãn này), và chọn nhầm nghĩa thì người dùng không có cách nào biết.
 */
export type TaskField =
  | 'Name'
  | 'BoardColumn'
  | 'Category'
  | 'Priority'
  | 'WorkItemType'
  | 'Sprint'
  | 'Assignee'
  | 'Reporter'
  | 'DueDate'
  | 'StoryPoints'
  | 'CreatedAt'
  /**
   * Trạng thái duyệt hiện thời (ADR-063) — giá trị là một `TaskApprovalState`.
   *
   * ⚠️ Trường dựng sẵn ĐẦU TIÊN không phải một cột của bảng Tasks: nó là phép chiếu của
   * bảng `Approvals` xuống task. Hệ quả ở UI: **không sắp xếp theo nó được** (backend cho
   * nó rơi về thứ tự mặc định), nên đừng bày nó ra ở ô "Sắp xếp theo".
   */
  | 'ApprovalState';

export const TASK_FIELD_LABEL: Record<TaskField, string> = {
  Name: 'Tên task',
  BoardColumn: 'Cột',
  Category: 'Nhóm trạng thái',
  Priority: 'Độ ưu tiên',
  WorkItemType: 'Loại công việc',
  Sprint: 'Sprint',
  Assignee: 'Người đảm nhận',
  Reporter: 'Người tạo',
  DueDate: 'Hạn hoàn thành',
  StoryPoints: 'Story Points',
  CreatedAt: 'Ngày tạo',
  ApprovalState: 'Trạng thái duyệt',
};

/**
 * Các trường SẮP XẾP được. `ApprovalState` cố ý vắng mặt — xem XML doc của nó.
 *
 * 🔴 Danh sách này tồn tại thay vì `Object.keys(TASK_FIELD_LABEL)`: bày ra một khoá sắp
 * xếp mà backend lặng lẽ bỏ qua là hứa với người dùng một thứ tự không tồn tại, và họ chỉ
 * phát hiện bằng cách nhìn kết quả và tự nghi ngờ mắt mình.
 */
export const SORTABLE_TASK_FIELDS: readonly TaskField[] = [
  'Name',
  'BoardColumn',
  'Category',
  'Priority',
  'WorkItemType',
  'Sprint',
  'Reporter',
  'DueDate',
  'StoryPoints',
  'CreatedAt',
];

/**
 * Trạng thái duyệt của một task, nhìn từ phía người lọc danh sách (ADR-063).
 *
 * ⚠️ KHÔNG phải `ApprovalStatus` (vòng đời của MỘT yêu cầu duyệt). Đây là phép chiếu của
 * các yêu cầu **còn hiệu lực** xuống chính task — hai enum không thay thế nhau được.
 *
 * 📌 Cố ý không có "chờ TÔI duyệt": một `SavedView` lưu giá trị **literal**, nên một view
 * CHIA SẺ mang điều kiện đó sẽ nói dối mọi người trừ tác giả của nó.
 */
export type TaskApprovalState = 'None' | 'Pending' | 'Approved' | 'Rejected';

export const TASK_APPROVAL_STATE_LABEL: Record<TaskApprovalState, string> = {
  None: 'Không cần duyệt',
  Pending: 'Đang chờ ký',
  Approved: 'Đã duyệt, chờ chuyển',
  Rejected: 'Bị từ chối',
};

/** Danh mục ĐÓNG. Toán tử nào hợp lệ với trường nào do `FilterValueKind` quyết định. */
export type FilterOperator =
  | 'Equals'
  | 'NotEquals'
  | 'Contains'
  | 'GreaterThan'
  | 'GreaterThanOrEqual'
  | 'LessThan'
  | 'LessThanOrEqual'
  | 'IsEmpty'
  | 'IsNotEmpty';

export const FILTER_OPERATOR_LABEL: Record<FilterOperator, string> = {
  Equals: 'bằng',
  NotEquals: 'khác',
  Contains: 'chứa',
  GreaterThan: 'lớn hơn',
  GreaterThanOrEqual: 'từ',
  LessThan: 'nhỏ hơn',
  LessThanOrEqual: 'đến',
  IsEmpty: 'chưa có',
  IsNotEmpty: 'đã có',
};

/** Hai toán tử duy nhất KHÔNG mang giá trị so sánh. */
export function isUnaryOperator(operator: FilterOperator): boolean {
  return operator === 'IsEmpty' || operator === 'IsNotEmpty';
}

/**
 * Hình dạng dữ liệu của một trường — quyết định toán tử nào hợp lệ và ô nhập nào được vẽ.
 *
 * 🔴 **Bản sao có chủ đích của `TaskFilterCatalog` ở backend.** Chấp nhận trùng vì nếu
 * không thì UI bày ra những toán tử backend sẽ từ chối, và người dùng chỉ biết sau khi bấm
 * Lưu. Backend vẫn là chốt chặn thật — đây chỉ là lớp không-bày-ra-thứ-sẽ-hỏng, đúng khuôn
 * `lib/tasks/permissions.ts` (bản sao của `ProjectPermissions.cs`).
 */
export type FilterValueKind = 'Text' | 'Number' | 'Date' | 'Boolean' | 'Reference' | 'Enum';

export function kindOfTaskField(field: TaskField): FilterValueKind {
  switch (field) {
    case 'Name':
      return 'Text';
    case 'BoardColumn':
    case 'WorkItemType':
    case 'Sprint':
    case 'Assignee':
    case 'Reporter':
      return 'Reference';
    case 'Category':
    case 'Priority':
    case 'ApprovalState':
      return 'Enum';
    case 'DueDate':
    case 'CreatedAt':
      return 'Date';
    case 'StoryPoints':
      return 'Number';
  }
}

export function kindOfFieldType(type: FieldType): FilterValueKind {
  switch (type) {
    case 'Text':
    case 'Url':
      return 'Text';
    case 'Number':
      return 'Number';
    case 'Date':
      return 'Date';
    case 'Checkbox':
      return 'Boolean';
    // Lọc theo Select là lọc theo LỰA CHỌN nào đang được chọn — giá trị là id của option,
    // không phải nhãn. So theo nhãn sẽ vỡ ngay lần đầu người dùng đổi tên một lựa chọn.
    case 'SingleSelect':
    case 'MultiSelect':
      return 'Reference';
  }
}

/** Toán tử hợp lệ với một kiểu giá trị — cùng luật với `TaskFilterCatalog.Supports`. */
export function operatorsFor(kind: FilterValueKind): FilterOperator[] {
  const base: FilterOperator[] = ['Equals', 'NotEquals', 'IsEmpty', 'IsNotEmpty'];

  if (kind === 'Text') return ['Equals', 'NotEquals', 'Contains', 'IsEmpty', 'IsNotEmpty'];

  // So sánh có thứ tự chỉ cho số và ngày. Enum KHÔNG có: `Priority.Highest = 0` nên
  // "lớn hơn Medium" trả về những việc ÍT ưu tiên hơn — đúng theo số, ngược hẳn với chữ.
  if (kind === 'Number' || kind === 'Date')
    return [
      'Equals',
      'NotEquals',
      'GreaterThan',
      'GreaterThanOrEqual',
      'LessThan',
      'LessThanOrEqual',
      'IsEmpty',
      'IsNotEmpty',
    ];

  return base;
}

/** Giá trị hợp lệ của `Category`/`Priority` — gửi lên dưới dạng TÊN, không phải số. */
export const CATEGORY_VALUES = ['ToDo', 'InProgress', 'Done'] as const;
export const PRIORITY_VALUES = ['Highest', 'High', 'Medium', 'Low', 'Lowest'] as const;

export interface SavedViewFilterDto {
  /** Loại trừ lẫn nhau với `fieldDefinitionId` — đúng một trong hai. */
  field: TaskField | null;
  fieldDefinitionId: string | null;
  operator: FilterOperator;
  /** `null` với `IsEmpty`/`IsNotEmpty`. Enum gửi TÊN, tham chiếu gửi id. */
  value: string | null;
}

export interface SavedViewColumnDto {
  field: TaskField | null;
  fieldDefinitionId: string | null;
}

export interface SavedViewResponse {
  id: string;
  projectId: string;
  name: string;
  ownerId: string;
  ownerName: string;
  isShared: boolean;
  order: number;
  sortBy: TaskField | null;
  sortDescending: boolean;
  groupBy: TaskField | null;
  filters: SavedViewFilterDto[];
  columns: SavedViewColumnDto[];
  /** Backend trả lời thay vì frontend tự suy — xem chú thích đầu file. */
  canEdit: boolean;
}

export interface CreateSavedViewRequest {
  name: string;
  isShared: boolean;
  sortBy: TaskField | null;
  sortDescending: boolean;
  groupBy: TaskField | null;
  filters?: SavedViewFilterDto[];
  columns?: SavedViewColumnDto[];
}

/** Ghi đè TOÀN PHẦN: `filters`/`columns` gửi lên thay thế hẳn cái cũ (ADR-044). */
export type UpdateSavedViewRequest = CreateSavedViewRequest;

export interface ReorderSavedViewsRequest {
  viewIds: string[];
}

/**
 * Đầu vào một lượt chạy. KHÔNG mang `viewId`: client gửi thẳng trạng thái view đang mở,
 * nên "xem thử trước khi lưu" dùng chung đúng đường này.
 */
export interface TaskQueryRequest {
  filters?: SavedViewFilterDto[];
  sortBy?: TaskField | null;
  sortDescending?: boolean;
  search?: string | null;
  page?: number;
  pageSize?: number;
}

export interface TaskListFieldValue {
  fieldDefinitionId: string;
  label: string;
  type: FieldType;
  valueText: string | null;
  valueNumber: number | null;
  valueDate: string | null;
  valueBoolean: boolean | null;
  selectedOptions: FieldOptionResponse[];
}

/**
 * Một dòng trên màn danh sách.
 *
 * ⚠️ `customFields` nằm NGOÀI `task` có chủ đích: `TaskSummaryResponse` nuôi cả board lẫn
 * backlog, và nhồi thêm vào đó sẽ buộc mọi query của hai màn kia phải nhớ include giá trị
 * trường — thiếu một chỗ là mảng rỗng một cách im lặng.
 */
export interface TaskListItemResponse {
  task: TaskSummaryResponse;
  customFields: TaskListFieldValue[];
}
